DICOM/src/Microsoft.Health.Dicom.Anonymizer.Core/Rules/AnonymizerRule.cs

```diff
@@ -0,0 +1,55 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using Dicom;
+using EnsureThat;
+using Microsoft.Health.Dicom.Anonymizer.Core.Exceptions;
+using Microsoft.Health.Dicom.Anonymizer.Core.Model;
+using Microsoft.Health.Dicom.Anonymizer.Core.Processors;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Dicom.Anonymizer.Core.Rules
+{
+    public abstract class AnonymizerRule
+    {
+        private readonly IAnonymizerProcessor _processor;
+
+        public AnonymizerRule(string method, string description, JObject ruleSetting = null, IAnonymizerProcessorFactory processorFactory = null, IAnonymizerSettingsFactory settingsFactory = null)
+        {
+            EnsureArg.IsNotNull(method, nameof(method));
+            EnsureArg.IsNotNull(description, nameof(description));
+
+            Description = description;
+            processorFactory ??= new DicomProcessorFactory();
+            _processor = processorFactory.CreateProcessor(method, ruleSetting, settingsFactory);
+        }
+
+        public string Description { get; set; }
+
+        public void Handle(DicomDataset dataset, ProcessContext context)
+        {
+            EnsureArg.IsNotNull(dataset, nameof(dataset));
+            EnsureArg.IsNotNull(context, nameof(context));
+
+            var locatedItems = LocateDicomTag(dataset, context);
+
+            foreach (var item in locatedItems)
+            {
+                if (_processor.IsSupportedVR(item))
+                {
+                    _processor.Process(dataset, item, context);
+                    context.VisitedNodes.Add(item.Tag.ToString());
+                }
+                else
+                {
+                    throw new AnonymizationOperationException(DicomAnonymizationErrorCode.UnsupportedAnonymizationFunction, $"Rule {Description} is not supported for the item with VR {item.ValueRepresentation}");
```

Comment:
>throw new AnonymizationOperationException [](http://example.com/codeflow?start=20&length=41)

do we really need to throw an exception here? My opinion is that we can simply ignore it and add some warnings in the log

Reply:
It is user's error in configuration file to set an unsupported method for item. Failing fast and letting customers correct the configuration file is a better experience?

Reply:
Looks like the overrided method LocateDicomTag filters items.
So is this (item not supported by the processor) a possible case or unexpected case? In what condition? need clarification

Reply:
LocateDicomTag locate items with matched tag. But not check the method is applicable for the item .
e.g. perturb function can not be used on any VR with string values.